src/sentry/static/sentry/app/views/organizationDiscover/aggregations/aggregation.jsx

```diff
@@ -0,0 +1,148 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {Box} from 'grid-emotion';
+import SelectControl from 'app/components/forms/selectControl';
+
+import {getInternal, getExternal} from './utils';
+import {TOPK_COUNTS} from '../data';
+
+export default class Aggregation extends React.Component {
+  static propTypes = {
+    value: PropTypes.array,
+    onChange: PropTypes.func,
+    columns: PropTypes.array,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      selectedFunction: null,
+    };
+  }
+
+  getOptions() {
+    const currentValue = getInternal(this.props.value);
+    return [{label: currentValue, value: currentValue}];
+  }
+
+  filterOptions = (options, input, value) => {
+    input = input || this.state.selectedFunction || '';
+
+    let optionList = [
+      {value: 'count', label: 'count'},
+      {value: 'uniq', label: 'uniq(...)'},
+      {value: 'topK', label: 'topK(...)'},
+      {value: 'avg', label: 'avg(...)'},
+    ];
+
+    if (input.startsWith('uniq')) {
+      optionList = this.props.columns.map(({name}) => ({
+        value: `uniq(${name})`,
+        label: `uniq(${name})`,
+      }));
+    }
+
+    if (input.startsWith('avg')) {
+      optionList = this.props.columns
+        .filter(({type}) => type === 'number')
+        .map(({name}) => ({
+          value: `avg(${name})`,
+          label: `avg(${name})`,
+        }));
+    }
+
+    if (input.startsWith('topK')) {
+      optionList = TOPK_COUNTS.map(num => ({
+        value: `topK(${num})`,
+        label: `topK(${num})(...)`,
+      }));
+    }
+
+    const topKValueMatch = input.match(/^topK\((\d+)\)/);
+
+    if (topKValueMatch) {
+      const count = topKValueMatch[1];
+
+      optionList = this.props.columns.map(({name}) => ({
+        value: `topK(${count})(${name})`,
+        label: `topK(${count})(${name})`,
+      }));
+    }
+
+    return optionList.filter(({label}) => label.includes(input));
+  };
+
+  focus() {
+    this.select.focus();
+  }
+
+  handleChange = option => {
+    const topKValues = new Set([...TOPK_COUNTS.map(num => `topK(${num})`)]);
+
+    if (option.value === 'uniq' || option.value === 'avg' || option.value === 'topK') {
+      this.setState({selectedFunction: option.value}, this.focus);
+    } else if (topKValues.has(option.value)) {
+      this.setState(
+        {
+          selectedFunction: option.value,
+        },
+        this.focus
+      );
+    } else {
+      this.setState({selectedFunction: null}, () => {
+        this.props.onChange(getExternal(option.value));
+      });
+    }
+  };
+
+  handleClose = () => {
+    this.setState({selectedFunction: null});
+  };
+
+  inputRenderer = props => {
+    const val = `${this.state.selectedFunction || ''}`.trim();
+
+    return (
+      <input
+        type="text"
+        {...props}
+        value={props.value || val}
+        style={{width: '100%', border: 0}}
```

Comment:
Can you make this a styled component?

Reply:
it's not super straightforward with react select, i'd come back and look closer at it later when tackling styling